File: frontend/src/services/sovereignty.ts
/**
 * Sovereignty API service
 */
import api from './api'

export interface SystemSovereignty {
  id: number
  system_id: number
  alliance_id: number | null
  corporation_id: number | null
  faction_id: number | null
  ihub_vulnerability_timer: number | null
  tcu_vulnerability_timer: number | null
  synced_at: string | null
}

export interface SovereigntyStructure {
  id: number
  structure_id: number
  system_id: number
  structure_type_id: number
  alliance_id: number
  vulnerable_start_time: string | null
  vulnerable_end_time: string | null
  vulnerability_occupancy_level: number | null
  synced_at: string | null
}

export interface SovereigntyCampaign {
  id: number
  campaign_id: number
  system_id: number
  constellation_id: number
  structure_id: number
  event_type: string
  defender_id: number | null
  defender_score: number
  attackers_score: number
  start_time: string
  synced_at: string | null
}

export interface SovereigntyStatistics {
  total_systems: number
  systems_by_alliance: { [key: string]: number }
  systems_by_faction: { [key: string]: number }
  vulnerable_structures: number
  active_campaigns: number
}

interface ListSystemsParams {
  alliance_id?: number
  corporation_id?: number
  faction_id?: number
  limit?: number
  offset?: number
}

interface ListStructuresParams {
  alliance_id?: number
  system_id?: number
  limit?: number
  offset?: number
}

interface ListCampaignsParams {
  system_id?: number
  defender_id?: number
  limit?: number
  offset?: number
}

export const sovereigntyService = {
  async listSystems(params: ListSystemsParams): Promise<SystemSovereignty[]> {
    const response = await api.get('/sov/systems', { params })
    return response.data
  },

  async getSystem(systemId: number): Promise<SystemSovereignty> {
    const response = await api.get(`/sov/systems/${systemId}`)
    return response.data
  },

  async listStructures(params: ListStructuresParams): Promise<SovereigntyStructure[]> {
    const response = await api.get('/sov/structures', { params })
    return response.data
  },

  async listCampaigns(params: ListCampaignsParams): Promise<SovereigntyCampaign[]> {
    const response = await api.get('/sov/campaigns', { params })
    return response.data
  },

  async getStatistics(): Promise<SovereigntyStatistics> {
    const response = await api.get('/sov/statistics')
    return response.data
  },

  async triggerSync(): Promise<void> {
    await api.post('/sov/sync')
  },
}

File: backend/app/models/sovereignty.py
"""
Sovereignty models for tracking system and constellation sovereignty
"""
from sqlalchemy import Column, Integer, BigInteger, String, Float, DateTime, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from app.core.database import Base


class SystemSovereignty(Base):
    """System sovereignty information"""
    __tablename__ = "system_sovereignty"

    id = Column(Integer, primary_key=True, index=True)
    system_id = Column(Integer, unique=True, nullable=False, index=True)

    # Ownership
    alliance_id = Column(Integer, index=True)
    corporation_id = Column(Integer, index=True)
    faction_id = Column(Integer, index=True)

    # Indexes
    ihub_vulnerability_timer = Column(Float)  # Activity Defense Multiplier
    tcu_vulnerability_timer = Column(Float)

    # Timestamps
    created_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()")
    updated_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()", onupdate="NOW()")
    synced_at = Column(DateTime(timezone=True))


class SovereigntyStructure(Base):
    """Sovereignty structures (TCU, IHUB)"""
    __tablename__ = "sovereignty_structures"

    id = Column(Integer, primary_key=True, index=True)
    structure_id = Column(BigInteger, unique=True, nullable=False, index=True)

    # Location
    system_id = Column(Integer, nullable=False, index=True)
    structure_type_id = Column(Integer, nullable=False)  # TCU or IHUB type

    # Ownership
    alliance_id = Column(Integer, nullable=False, index=True)

    # Vulnerability
    vulnerable_start_time = Column(DateTime(timezone=True))
    vulnerable_end_time = Column(DateTime(timezone=True))
    vulnerability_occupancy_level = Column(Float)

    # Timestamps
    created_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()")
    updated_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()", onupdate="NOW()")
    synced_at = Column(DateTime(timezone=True))


class SovereigntyCampaign(Base):
    """Ongoing sovereignty campaigns (contests)"""
    __tablename__ = "sovereignty_campaigns"

    id = Column(Integer, primary_key=True, index=True)
    campaign_id = Column(Integer, unique=True, nullable=False, index=True)

    # Location
    system_id = Column(Integer, nullable=False, index=True)
    constellation_id = Column(Integer, nullable=False, index=True)

    # Structure under attack
    structure_id = Column(BigInteger, nullable=False)
    event_type = Column(String(100), nullable=False)  # tcu_defense, ihub_defense, station_defense

    # Participants
    defender_id = Column(Integer, index=True)  # Alliance ID
    defender_score = Column(Float, default=0)

    attackers_score = Column(Float, default=0)

    # Timing
    start_time = Column(DateTime(timezone=True), nullable=False)

    # Timestamps
    created_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()")
    updated_at = Column(DateTime(timezone=True), nullable=False, server_default="NOW()", onupdate="NOW()")
    synced_at = Column(DateTime(timezone=True))
